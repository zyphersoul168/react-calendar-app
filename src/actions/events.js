import Swal from "sweetalert2";
import { types } from "../actionTypes/types";
import { fetchWithToken } from "../helpers/fetch";
import { parseEvents } from "../helpers/parseEvents";

// async actions
export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;

    try {
      const resp = await fetchWithToken("events", event, "POST");
      const body = await resp.json();

      if (body.ok) {
        event._id = body.event._id;
        event.user = {
          _id: uid,
          name: name,
        };
        console.log(event);
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken("events");
      const body = await resp.json();
      const events = parseEvents(body.events);
      dispatch(eventLoad(events));
      
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartUpdate = (event) => {
  return async(dispatch) => {
    try {
      const resp =  await fetchWithToken(`events/${event._id}`, event, "PUT");
      const body = await resp.json();

      if(body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  }
}

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    const { _id } = getState().calendar.activeEvent;

    try {
      const resp =  await fetchWithToken(`events/${_id}`, {}, "DELETE");
      const body = await resp.json();

      if(body.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  }
}

// sync actions
export const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActive = () => ({
  type: types.eventClearActive,
});

export const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event,
});

export const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventLoad = (events) => ({
  type: types.eventLoad,
  payload: events,
}) 

export const eventLogout = () => ({
  type: types.eventLogout,
});
