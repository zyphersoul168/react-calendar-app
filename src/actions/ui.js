import { types } from "../actionTypes/types";

export const uiOpenModal = () => ({
  type: types.uiOpenModal,
});

export const uiCloseModal = () => ({
  type: types.uiCloseModal,
});
