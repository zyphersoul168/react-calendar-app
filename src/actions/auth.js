import Swal from "sweetalert2";
import { types } from "../actionTypes/types";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const response = await fetchWithoutToken(
      "auth/login",
      { email, password },
      "POST"
    );
    const body = await response.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startSignup = (name, email, password) => {
  return async (dispatch) => {
    const response = await fetchWithoutToken(
      "auth/signup",
      { name, email, password },
      "POST",
    );

    const body = await response.json();

    if(body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  }
};

export const startChecking = () => {
  return async (dispatch) => {
    const response = await fetchWithToken("auth/resignup");

    const body = await response.json();

    if(body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      dispatch(setCheckingFinish());
    }
  }
}

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(eventLogout());
    dispatch(logout());
  };
};

// sync actions
export const setCheckingFinish = () => ({
  type: types.authCheckingFinish,
});

export const login = (user) => ({
  type: types.authLogin,
  payload: user,
}); 

export const logout = () => ({
  type: types.authLogout,
});