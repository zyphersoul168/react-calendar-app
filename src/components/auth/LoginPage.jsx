import React from "react";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { startLogin } from "../../actions/auth";
import { useForm } from "../../hooks/useForm";

export const LoginPage = () => {
  const [formValues, handleInputChange] = useForm({
    email: "",
    password: "",
  });

  const dispatch = useDispatch();

  const { email, password } = formValues;

  const handleLogin = (event) => {
    event.preventDefault();

    dispatch(startLogin(email, password));
  }

  return (
    <div className="flex relative items-center right-16">
      <div className="absolute -right-32 z-10 space-y-8 bg-gray-100 border border-dark-lava p-10 rounded-lg shadow-lg text-center">
        <h1 className="text-dark-lava font-abril text-6xl">
          Log <span className="text-melon">in</span>
        </h1>
        {/* <hr className="bg-dark-lava border-0 h-0.5" /> */}
        <div>
          <form onSubmit={handleLogin} className="space-y-3">
            <input
              className="block p-2 bg-gray-200 focus:outline-none placeholder-middle-gray text-dark-lava rounded-md w-full"
              type="email"
              name="email"
              placeholder="Email"
              value={email}
              onChange={handleInputChange}
            />
            <input
              className="block p-2 bg-gray-200 focus:outline-none placeholder-middle-gray text-dark-lava rounded-md w-full"
              type="password"
              name="password"
              placeholder="Password"
              value={password}
              onChange={handleInputChange}
            />
            <button
              className="block font-bold p-2 bg-melon text-dark-lava rounded-md w-full"
              type="submit"
            >
              Log In
            </button>
          </form>
          <div className="mt-2">
            <p className="text-dark-lava">
              Don't have an account?{" "}
              <Link to="/auth/signup">
                <span className="text-melon font-bold">Sign up</span>
              </Link>
            </p>
          </div>
        </div>
      </div>
      <div className="relative z-0">
        <img
          className="border border-dark-lava rounded-2xl"
          src="https://cdn.pixabay.com/photo/2021/07/25/16/13/plants-6492284_960_720.jpg"
          alt="plants"
        />
      </div>
    </div>
  );
};
