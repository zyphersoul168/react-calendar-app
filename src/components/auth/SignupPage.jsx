import React from 'react'
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom'
import Swal from 'sweetalert2';
import { startSignup } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const SignupPage = () => {
  const dispatch = useDispatch();
  
  const [formValues, handleInputChange] = useForm({
    name: "",
    email: "",
    password: "",
    password2: ""
  });

  const {name, email, password, password2} = formValues;
  
  const handleSignup = (event) => {
    event.preventDefault();

    if(password.length < 5) {
      return Swal.fire("Error", "Password must have at least 6 chars", "error");
    }
    else if (password !== password2) {
      return Swal.fire("Error", "Password must match", "error");
    }

    dispatch(startSignup(name, email, password));
  };

  return (
    <div className="flex flex-row-reverse relative items-center left-16">
      <div className="absolute lg:-left-32 xs:-left-0 z-10 space-y-8 bg-gray-100 border border-dark-lava p-10 rounded-lg shadow-lg">
        <h1 className="text-gray-800 font-abril text-6xl">
          Sign <span className="text-purple-500">up</span>
        </h1>
        {/* <hr className="bg-dark-lava border-0 h-0.5" /> */}
        <div>
          <form onSubmit={handleSignup} className="space-y-3">
            <input
              className="block p-2 bg-gray-200 focus:outline-none placeholder-gray-400 text-gray-700 rounded-md w-full"
              type="text"
              name="name"
              placeholder="Name"
              value={name}
              onChange={handleInputChange}
            />
            <input
              className="block p-2 bg-gray-200 focus:outline-none placeholder-gray-400 text-gray-700 rounded-md w-full"
              type="email"
              name="email"
              placeholder="Email"
              value={email}
              onChange={handleInputChange}
            />
            <input
              className="block p-2 bg-gray-200 focus:outline-none placeholder-gray-400 text-gray-700 rounded-md w-full"
              type="password"
              name="password"
              placeholder="Password"
              value={password}
              onChange={handleInputChange}
            />
            <input
              className="block p-2 bg-gray-200 focus:outline-none placeholder-gray-400 text-gray-700 rounded-md w-full"
              type="password"
              name="password2"
              placeholder="Confirm password"
              value={password2}
              onChange={handleInputChange}
            />
            <button
              className="block font-bold p-2 bg-yellow-200 hover:bg-yellow-300 text-gray-800 w-full rounded-md transition ease-in"
              type="submit"
            >
              Sign up
            </button>
          </form>
          <div className="mt-2">
            <p className="text-gray-800">
              Already have an account?{" "}
              <Link to="/auth/login">
                <span className="text-pink-400 font-bold">Log in</span>
              </Link>
            </p>
          </div>
        </div>
      </div>
      <div className="relative z-0">
        <img
          className="border border-dark-lava rounded-2xl"
          src="https://cdn.pixabay.com/photo/2021/04/25/15/41/flower-6206819_960_720.jpg"
          alt="plants"
        />
      </div>
    </div>
  )
}
