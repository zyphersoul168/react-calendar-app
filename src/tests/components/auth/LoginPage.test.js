import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom";

import { LoginPage } from "../../../components/auth/LoginPage";
import { startLogin } from "../../../actions/auth";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {};
const store = mockStore(initialState);

// mocks
store.dispatch = jest.fn();
jest.mock("../../../actions/auth", () => ({
  startLogin: jest.fn(),
}));

const wrapper = mount(
  <Provider store={store}>
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
  </Provider>
);

describe("Tests on LoginPage component", () => {
  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should dispatch login action', () => {
    wrapper.find('input[name="email"]').simulate("change", {
      target: {
        name: "email",
        value: "aaron@gmail.com",
      },
    });

    wrapper.find('input[name="password"]').simulate("change", {
      target: {
        name: "password",
        value: "123456",
      },
    });

    wrapper.find("form").prop("onSubmit")({
      preventDefault(){},
    });
    
    expect(startLogin).toHaveBeenCalledWith("aaron@gmail.com", "123456");
  });
  
});
