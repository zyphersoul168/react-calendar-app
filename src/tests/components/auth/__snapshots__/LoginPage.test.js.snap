// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on LoginPage component should match snapshot 1`] = `
<div
  className="flex relative items-center right-16"
>
  <div
    className="absolute -right-32 z-10 space-y-8 bg-gray-100 border border-dark-lava p-10 rounded-lg shadow-lg text-center"
  >
    <h1
      className="text-dark-lava font-abril text-6xl"
    >
      Log 
      <span
        className="text-melon"
      >
        in
      </span>
    </h1>
    <div>
      <form
        className="space-y-3"
        onSubmit={[Function]}
      >
        <input
          className="block p-2 bg-gray-200 focus:outline-none placeholder-middle-gray text-dark-lava rounded-md w-full"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          className="block p-2 bg-gray-200 focus:outline-none placeholder-middle-gray text-dark-lava rounded-md w-full"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          className="block font-bold p-2 bg-melon text-dark-lava rounded-md w-full"
          type="submit"
        >
          Log In
        </button>
      </form>
      <div
        className="mt-2"
      >
        <p
          className="text-dark-lava"
        >
          Don't have an account?
           
          <a
            href="/auth/signup"
            onClick={[Function]}
          >
            <span
              className="text-melon font-bold"
            >
              Sign up
            </span>
          </a>
        </p>
      </div>
    </div>
  </div>
  <div
    className="relative z-0"
  >
    <img
      alt="plants"
      className="border border-dark-lava rounded-2xl"
      src="https://cdn.pixabay.com/photo/2021/07/25/16/13/plants-6492284_960_720.jpg"
    />
  </div>
</div>
`;
